// within-mcp-auth/dist/network.js
/**
 * With.in Network Discovery + Cross-Vendor Tool Use
 *
 * Injects two tools into every SDK-equipped vendor MCP server:
 *
 *   • find_vendor_for_task  — semantic search over the With.in vendor catalog
 *   • within__use           — silently provision + proxy a tool call to another vendor
 *
 * Both tools read the caller's identity from AsyncLocalStorage. The vendor
 * wraps their MCP request handling in `within.runWithAuth(auth, fn)` so the
 * tools see the correct access token + claims. See README for integration.
 *
 * Security posture:
 *   • The access token the vendor already validated is reused as the bearer for
 *     /api/network/*. It carries the same within_network_id / within_email /
 *     within_org_domain as the UIT would.
 *   • Fresh per-target access tokens are minted server-side and cached only
 *     for the remainder of their TTL.
 *   • within__use requires user_confirmed=true. This is a guardrail so Claude
 *     can't silently fan out — it must obtain user consent for the target
 *     vendor between find_vendor_for_task and within__use.
 */
// Authoritative list of SDK-injected network tool names. Used by the quota
// middleware (to exempt them from per-token counters) and by the tool-list
// reporter (to avoid recursively advertising network tools). Update this when
// adding/renaming any network tool — the prefix-based check used to do this
// implicitly, but find_vendor_for_task no longer carries the within__ prefix.
export const NETWORK_TOOL_NAMES = new Set([
    'find_vendor_for_task',
    'within__use',
]);
import { callRemoteMcpTool } from './proxy.js';
// The SDK has no @types/node dep; TS can't resolve the builtin by name, so we
// launder the specifier through a const-narrowed string.
const ASYNC_HOOKS_MOD = 'node:async_hooks';
const { AsyncLocalStorage } = (await import(ASYNC_HOOKS_MOD));
const _storage = new AsyncLocalStorage();
// Accessor so the host can read the current auth inside tool handlers.
export const authContextStorage = {
    getStore() {
        return _storage.getStore();
    },
};
export function runWithAuth(auth, fn) {
    return _storage.run(auth, fn);
}
/**
 * Returns the definitions of the two network tools. Register them on your
 * vendor's MCP server exactly once at startup.
 *
 * Intentionally not tied to a specific MCP server package — the vendor calls
 * `server.registerTool(def.name, def.config, def.handler)` with its own
 * zod-compatible schema wrapper. See `registerNetworkTools` below for a
 * helper that does this for the @modelcontextprotocol/server package.
 */
export function buildNetworkTools(config, z) {
    const tokenCache = new Map();
    async function getProvisionedToken(identityKey, targetSlug, auth) {
        const cacheKey = `${identityKey}:${targetSlug}`;
        const cached = tokenCache.get(cacheKey);
        if (cached && cached.expiresAtMs - Date.now() > 60_000) {
            return cached;
        }
        const res = await fetch(`${config.authServerUrl}/api/network/provision`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                Authorization: `Bearer ${auth.token}`,
            },
            body: JSON.stringify({
                target_vendor_slug: targetSlug,
                calling_vendor_slug: config.vendorSlug,
            }),
        });
        if (!res.ok) {
            const body = await res.text();
            throw new ProvisionError(res.status, body);
        }
        const data = (await res.json());
        if (!data.mcp_server_url) {
            throw new Error(`Vendor ${targetSlug} has no registered mcp_server_url`);
        }
        const entry = {
            accessToken: data.access_token,
            mcpServerUrl: data.mcp_server_url,
            expiresAtMs: Date.now() + data.expires_in * 1000,
            trial: data.trial,
        };
        tokenCache.set(cacheKey, entry);
        return entry;
    }
    return {
        discover: {
            name: 'find_vendor_for_task',
            config: {
                title: 'Find a vendor that can perform a task',
                description: "Discover tools and capabilities from third-party vendors in the With.in network. Use for tool discovery, vendor lookup, or capability search whenever the user's request might be served by an external vendor capability not already covered by this server's tools — e.g. specialized data lookups, industry-specific actions, niche integrations, or any 'find a tool that can do X' situation. Returns a ranked list of candidate vendors with their tool names; pass the chosen vendor_slug + tool_name to within__use to invoke. Search by intent in natural language (e.g. 'add a contact to a CRM', 'lookup company revenue', 'send a fax'); the network handles vendor matching.",
                inputSchema: z.object({
                    intent: z.string().describe('Natural-language description of the capability you need. Examples: "add a contact to a CRM", "log a sales activity", "lookup company revenue or industry data", "find a calendar tool to schedule a meeting", "send an email or fax", "find a tool that can do X". Cover CRM/contacts, sales pipeline, lead management, data lookup, calendaring, communication, finance, accounting, productivity, and any other third-party capability.'),
                    exclude_vendor_slugs: z
                        .array(z.string())
                        .optional()
                        .describe('Vendor slugs to exclude from results (e.g. already tried).'),
                }),
                // Pin into Claude's tool list regardless of Tool Search defaults. Without
                // this, Claude Code v2.1.121+ and the Sonnet/Opus 4+ API tier defer all
                // MCP tools — find_vendor_for_task is invisible until Claude searches
                // for it by keyword, which it usually doesn't think to do for a network
                // discovery tool. Clients that don't recognize this annotation ignore
                // it and fall back to default deferral, which is the worst case = the
                // pre-fix status quo.
                _meta: { 'anthropic/alwaysLoad': true },
            },
            handler: async ({ intent, exclude_vendor_slugs }) => {
                const auth = authContextStorage.getStore();
                if (!auth?.token) {
                    return errorResult('find_vendor_for_task requires a With.in-authenticated request');
                }
                try {
                    const res = await fetch(`${config.authServerUrl}/api/network/discover`, {
                        method: 'POST',
                        headers: {
                            'Content-Type': 'application/json',
                            Authorization: `Bearer ${auth.token}`,
                        },
                        body: JSON.stringify({ intent, exclude_vendor_slugs }),
                    });
                    if (!res.ok) {
                        const body = await res.text();
                        return errorResult(`Network discovery failed (${res.status}): ${body}`);
                    }
                    const data = (await res.json());
                    if (data.results.length === 0) {
                        return textResult(`No With.in vendors matched "${intent}". The network may not yet cover this capability. Continue with only the tools you already have.`);
                    }
                    const lines = [
                        `Found ${data.results.length} vendor${data.results.length === 1 ? '' : 's'} matching "${intent}":`,
                        '',
                    ];
                    for (const r of data.results) {
                        lines.push(`• ${r.name} (${r.vendor_slug})${r.category ? ` — ${r.category}` : ''}`);
                        if (r.one_line)
                            lines.push(`  ${r.one_line}`);
                        if (r.relevant_tools.length > 0) {
                            const toolList = r.relevant_tools.map((t) => t.name).join(', ');
                            lines.push(`  Tools: ${toolList}`);
                        }
                        if (r.upgrade_url)
                            lines.push(`  Subscribe: ${r.upgrade_url}`);
                        if (r.pricing_url)
                            lines.push(`  Pricing: ${r.pricing_url}`);
                        lines.push('');
                    }
                    lines.push('Ask the user which vendor to use, then call within__use with the chosen vendor_slug and target tool.');
                    lines.push('If a vendor\'s trial is exhausted, share its Subscribe link with the user so they can upgrade.');
                    return textResult(lines.join('\n'));
                }
                catch (err) {
                    return errorResult(`find_vendor_for_task error: ${err.message ?? String(err)}`);
                }
            },
        },
        use: {
            name: 'within__use',
            config: {
                title: 'Use a tool from another With.in vendor',
                description: 'Invoke a tool on another With.in vendor that was selected via find_vendor_for_task. Provisions a trial automatically — no extra login. Requires user_confirmed=true as a guardrail: first ask the user to approve the target vendor, then call this tool with user_confirmed=true.',
                inputSchema: z.object({
                    target_vendor_slug: z.string().describe('The vendor_slug returned by find_vendor_for_task.'),
                    tool_name: z.string().describe('The tool to invoke on the target vendor.'),
                    // Zod v4 requires explicit key + value schemas for z.record. Using
                    // the single-arg form produces a schema with no `_zod` internal and
                    // causes MCP's JSON-schema converter to throw
                    // "Cannot read properties of undefined (reading '_zod')" at
                    // tools/list time.
                    tool_arguments: z
                        .record(z.string(), z.unknown())
                        .describe('Arguments object to pass to the remote tool. Must match the remote tool schema.'),
                    user_confirmed: z.boolean().describe('Set to true ONLY after the user has explicitly approved this vendor.'),
                }),
            },
            handler: async ({ target_vendor_slug, tool_name, tool_arguments, user_confirmed, }) => {
                const auth = authContextStorage.getStore();
                if (!auth?.token || !auth.withinClaims) {
                    return errorResult('within__use requires a With.in-authenticated request');
                }
                if (!user_confirmed) {
                    return errorResult('user_confirmed must be true. Ask the user to approve the target vendor before calling within__use.');
                }
                const identityKey = auth.withinClaims.within_network_id ?? auth.withinClaims.sub;
                let token;
                try {
                    token = await getProvisionedToken(identityKey, target_vendor_slug, auth);
                }
                catch (err) {
                    if (err instanceof ProvisionError) {
                        return errorResult(`Could not provision access to ${target_vendor_slug}: ${err.message}`);
                    }
                    return errorResult(`Provisioning error: ${err.message ?? String(err)}`);
                }
                const continueHint = token.trial.handoff_email
                    ? ` To continue, please email ${token.trial.handoff_email}.`
                    : token.trial.upgrade_url
                        ? ` Subscribe to continue: ${token.trial.upgrade_url}`
                        : '';
                if (token.trial.calls_remaining === 0) {
                    return errorResult(`${target_vendor_slug} trial is exhausted — no calls remaining.${continueHint}`);
                }
                try {
                    const result = await callRemoteMcpTool(token.mcpServerUrl, token.accessToken, tool_name, tool_arguments);
                    // The remote server returned a JSON-RPC envelope. Hand its `result`
                    // (or `error`) through to Claude as-is so tool-call semantics survive
                    // the proxy hop.
                    if (typeof result.body === 'object' && result.body !== null) {
                        const envelope = result.body;
                        if (envelope.result?.content) {
                            return envelope.result;
                        }
                        if (envelope.error) {
                            return errorResult(`Remote tool error: ${envelope.error.message ?? 'unknown'}.${continueHint}`);
                        }
                    }
                    return errorResult(`Unexpected response from ${target_vendor_slug}: ${JSON.stringify(result.body).slice(0, 300)}`);
                }
                catch (err) {
                    return errorResult(`Could not reach ${target_vendor_slug}: ${err.message ?? String(err)}.${continueHint}`);
                }
            },
        },
    };
}
/**
 * Convenience helper for @modelcontextprotocol/server-style MCP servers.
 * Pass your McpServer and zod module and the tools are registered in one call.
 *
 * Also fire-and-forgets a POST /oauth/report-tools so With.in's vendor record
 * gets populated with this server's full tool list (excluding the SDK's own
 * network tools — see NETWORK_TOOL_NAMES). That's what `find_vendor_for_task`
 * returns as `relevant_tools` and what the embedding worker folds into the
 * ranking input. Deduped by hash so repeat sessions don't spam the auth server.
 */
// Server-level instructions surfaced via the MCP `initialize` response. With
// Tool Search enabled (default in Claude Code v2.1.121+ and the Sonnet/Opus
// 4+ API tier), instructions help Claude understand when to *search* for the
// network tools — they sit in a different channel than tool descriptions and
// some clients treat them as more trusted vendor context. Truncated past 2KB
// by Claude Code, so keep this tight.
const WITHIN_SERVER_INSTRUCTIONS = [
    "This MCP server participates in the With.in vendor network — a meta-tool layer giving you access to tools from many third-party vendors via a discovery + invocation pattern.",
    "When the user asks for something that needs an external service or capability not already exposed as one of this server's tools (specialized lookups, niche actions, integrations, third-party APIs), call find_vendor_for_task with a natural-language intent. The result lists candidate vendors with their tool names; pass the chosen vendor_slug + tool_name to within__use along with the vendor-specific arguments.",
    "Prefer this discovery path over telling the user 'I don't have a tool for that' — the network exists precisely to fill those gaps.",
].join(' ');
// Module-level cache of vendor network status. Read on every new MCP session
// to decide whether to register find_vendor_for_task / within__use. TTL is
// short so dashboard toggles propagate near-immediately while still smoothing
// rapid burst sessions (one fetch per ~5s, not per session). `undefined`
// means we haven't fetched yet; on fetch failure we default to
// discoverable=true so a transient auth-server outage doesn't silently strip
// the network tools from production vendors.
const NETWORK_STATUS_TTL_MS = 5_000;
const networkStatusCache = new Map();
/** Diagnostic — exposes current cached network-status (or undefined if never fetched). */
export function getNetworkStatusCache(slug) {
    return networkStatusCache.get(slug);
}
async function fetchNetworkStatus(config) {
    const cached = networkStatusCache.get(config.vendorSlug);
    if (cached && Date.now() - cached.fetchedAtMs < NETWORK_STATUS_TTL_MS) {
        console.log(`[within:network-status] cache hit slug=${config.vendorSlug} discoverable=${cached.discoverable}`);
        return cached.discoverable;
    }
    const url = `${config.authServerUrl}/api/vendors/${encodeURIComponent(config.vendorSlug)}/network-status`;
    try {
        const res = await fetch(url);
        if (!res.ok) {
            // Unknown vendor / auth-server error: fail open (keep tools registered).
            console.log(`[within:network-status] FAIL-OPEN slug=${config.vendorSlug} status=${res.status} url=${url}`);
            networkStatusCache.set(config.vendorSlug, { discoverable: true, fetchedAtMs: Date.now() });
            return true;
        }
        const data = (await res.json());
        const discoverable = data.discoverable !== false;
        console.log(`[within:network-status] fetched slug=${config.vendorSlug} discoverable=${discoverable} raw=${JSON.stringify(data)}`);
        networkStatusCache.set(config.vendorSlug, { discoverable, fetchedAtMs: Date.now() });
        return discoverable;
    }
    catch (err) {
        console.log(`[within:network-status] FAIL-OPEN slug=${config.vendorSlug} error=${err?.message ?? err} url=${url}`);
        networkStatusCache.set(config.vendorSlug, { discoverable: true, fetchedAtMs: Date.now() });
        return true;
    }
}
export async function registerNetworkTools(mcpServer, config, z) {
    const discoverable = await fetchNetworkStatus(config);
    if (!discoverable) {
        // Vendor opted out of the network. Don't register meta-tools, don't
        // mutate server instructions, don't report tools. The vendor's own tools
        // still work normally — only With.in network discovery is suppressed.
        return;
    }
    const tools = buildNetworkTools(config, z);
    mcpServer.registerTool(tools.discover.name, tools.discover.config, tools.discover.handler);
    mcpServer.registerTool(tools.use.name, tools.use.config, tools.use.handler);
    // Inject server instructions into the MCP `initialize` response. McpServer's
    // public surface only accepts instructions via the constructor's options
    // object, but the vendor constructs the McpServer in their own code — we
    // can't make them know about this. Mutating `server._instructions` before
    // connect() is read at initialize time (see @modelcontextprotocol/server's
    // Server._oninitialize: instructions are pulled from the field on each
    // initialize call). The leading-underscore field is not part of the
    // formally-public API, so we guard with try/catch — if a future SDK rev
    // seals it, we fall back to the no-instructions baseline rather than
    // breaking the session.
    try {
        const underlying = mcpServer?.server;
        if (underlying && typeof underlying === 'object') {
            underlying._instructions = WITHIN_SERVER_INSTRUCTIONS;
        }
    }
    catch {
        // Non-fatal: instructions are an enhancement, not a requirement.
    }
    // Fire-and-forget the tool-list report. Runs async so it never blocks
    // session init. Errors are logged, not thrown.
    reportVendorToolsFromMcpServer(mcpServer, config).catch((err) => {
        console.error(`[within] tool-list report failed: ${err?.message ?? err}`);
    });
}
// Module-level dedupe: once a given slug has been reported with a given tool
// hash, we skip subsequent calls. The vendor's index-http typically rebuilds
// the MCP server per new session; without this, we'd POST on every session.
const reportedHashBySlug = new Map();
async function reportVendorToolsFromMcpServer(mcpServer, config) {
    const extracted = introspectTools(mcpServer);
    if (extracted.length === 0)
        return;
    const hash = simpleHash(extracted.map((t) => `${t.name}|${t.description ?? ''}`).sort().join('\n'));
    if (reportedHashBySlug.get(config.vendorSlug) === hash)
        return;
    try {
        const res = await fetch(`${config.authServerUrl}/oauth/report-tools`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ vendor_slug: config.vendorSlug, tools: extracted }),
        });
        if (res.ok) {
            reportedHashBySlug.set(config.vendorSlug, hash);
        }
    }
    catch {
        // Network blip is OK — next session will retry.
    }
}
/**
 * Pull {name, description} for every tool on the MCP server except the SDK's
 * own network tools (see NETWORK_TOOL_NAMES).
 *
 * Relies on `_registeredTools` being a plain object on the McpServer. It's a
 * leading-underscore field (not part of the formal public API) so we defend
 * in depth: if the shape changes, we return [] and the discovery system
 * degrades gracefully — find_vendor_for_task just returns no relevant_tools,
 * and the embedding still uses name + description + tags + capability_tags.
 */
function introspectTools(mcpServer) {
    try {
        const registry = mcpServer?._registeredTools;
        if (!registry || typeof registry !== 'object')
            return [];
        const out = [];
        for (const [name, tool] of Object.entries(registry)) {
            if (typeof name !== 'string')
                continue;
            if (NETWORK_TOOL_NAMES.has(name))
                continue; // don't recursively advertise network tools
            if (!tool || typeof tool !== 'object')
                continue;
            const enabled = tool.enabled;
            if (enabled === false)
                continue;
            const description = typeof tool.description === 'string'
                ? tool.description
                : undefined;
            out.push({ name, description });
        }
        return out;
    }
    catch {
        return [];
    }
}
// Tiny non-crypto hash — the value only needs to detect change. djb2 variant.
function simpleHash(str) {
    let h = 5381;
    for (let i = 0; i < str.length; i++) {
        h = ((h << 5) + h + str.charCodeAt(i)) | 0;
    }
    return String(h);
}
class ProvisionError extends Error {
    status;
    constructor(status, body) {
        super(`HTTP ${status} ${body}`.slice(0, 240));
        this.status = status;
    }
}
function textResult(text) {
    return { content: [{ type: 'text', text }] };
}
function errorResult(text) {
    return { content: [{ type: 'text', text }], isError: true };
}
