// within-mcp-auth/dist/proxy.js
/**
 * Remote MCP tool-call proxy — Streamable HTTP.
 *
 * Used by within__use to invoke a tool on another vendor's MCP server.
 *
 * The MCP Streamable HTTP transport requires a handshake before tool calls:
 *   1. POST /mcp  { method: "initialize", ... }        → 200 with mcp-session-id header
 *   2. POST /mcp  { method: "notifications/initialized" } (with that session id)
 *   3. POST /mcp  { method: "tools/call", ... }        (with that session id) → result
 *
 * Skipping steps 1+2 yields "Server not initialized" on the target server,
 * which is what killed within__use against the real Beacon CRM deployment.
 *
 * Response bodies come back as either application/json (a single JSON-RPC
 * response) or text/event-stream (SSE with one or more `event: message \n
 * data: {...}` blocks). We accept both and extract the first JSON-RPC
 * response message we see.
 *
 * Session reuse: we cache the session id per (url, token) so a subsequent
 * within__use to the same target doesn't re-handshake. Cache keyed on access
 * token which itself has a short TTL, so there's no long-lived drift risk.
 */
const sessionCache = new Map();
const SESSION_TTL_MS = 9 * 60_000; // generous buffer under the 10-min access-token TTL
const PROTOCOL_VERSION = '2025-11-25';
export async function callRemoteMcpTool(mcpServerUrl, accessToken, toolName, toolArguments, options) {
    const timeoutMs = options?.timeoutMs ?? 30_000;
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeoutMs);
    try {
        const sessionId = await ensureSession(mcpServerUrl, accessToken, controller.signal);
        const res = await fetch(mcpServerUrl, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                Accept: 'application/json, text/event-stream',
                Authorization: `Bearer ${accessToken}`,
                'mcp-session-id': sessionId,
            },
            body: JSON.stringify({
                jsonrpc: '2.0',
                id: options?.requestId ?? 1,
                method: 'tools/call',
                params: { name: toolName, arguments: toolArguments },
            }),
            signal: controller.signal,
        });
        const text = await res.text();
        const parsed = parseJsonRpcResponse(text, res.headers.get('content-type') ?? '');
        // If the target says our session has vanished (container recycled,
        // memory eviction, etc.) drop our cache entry and retry exactly once.
        if (!res.ok && looksLikeStaleSession(text)) {
            sessionCache.delete(cacheKey(mcpServerUrl, accessToken));
            return callRemoteMcpTool(mcpServerUrl, accessToken, toolName, toolArguments, options);
        }
        return { ok: res.ok, status: res.status, body: parsed ?? text };
    }
    finally {
        clearTimeout(timer);
    }
}
function cacheKey(url, token) {
    return `${url}::${token}`;
}
function looksLikeStaleSession(body) {
    return /session\s+(expired|not\s+found|invalid)/i.test(body) || body.includes('Session expired');
}
async function ensureSession(url, accessToken, signal) {
    const key = cacheKey(url, accessToken);
    const cached = sessionCache.get(key);
    if (cached && Date.now() - cached.createdAt < SESSION_TTL_MS) {
        return cached.sessionId;
    }
    const initRes = await fetch(url, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            Accept: 'application/json, text/event-stream',
            Authorization: `Bearer ${accessToken}`,
        },
        body: JSON.stringify({
            jsonrpc: '2.0',
            id: 0,
            method: 'initialize',
            params: {
                protocolVersion: PROTOCOL_VERSION,
                capabilities: {},
                clientInfo: { name: 'within-mcp-auth/proxy', version: '0.1.0' },
            },
        }),
        signal,
    });
    if (!initRes.ok) {
        const errText = await initRes.text();
        throw new Error(`Remote MCP initialize failed (${initRes.status}): ${errText.slice(0, 240)}`);
    }
    const sessionId = initRes.headers.get('mcp-session-id');
    if (!sessionId) {
        throw new Error('Remote MCP server did not return an mcp-session-id header on initialize');
    }
    // Consume init response body to free the socket before the next POST.
    await initRes.text().catch(() => undefined);
    // Fire-and-forget the initialized notification. The spec wants it, and
    // some servers block tools/call until they've seen it.
    const notifRes = await fetch(url, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            Accept: 'application/json, text/event-stream',
            Authorization: `Bearer ${accessToken}`,
            'mcp-session-id': sessionId,
        },
        body: JSON.stringify({
            jsonrpc: '2.0',
            method: 'notifications/initialized',
        }),
        signal,
    });
    // 202 Accepted is the expected response for notifications; anything else
    // we tolerate (target may acknowledge with 200 or drop it). We do NOT
    // throw here because some implementations don't require this step.
    await notifRes.text().catch(() => undefined);
    sessionCache.set(key, { sessionId, createdAt: Date.now() });
    return sessionId;
}
/**
 * The MCP StreamableHTTP transport can respond with either raw JSON or SSE.
 * This function peels off the SSE framing (if present) and returns the
 * first JSON-RPC response object it finds, or null if the body is neither.
 */
function parseJsonRpcResponse(text, contentType) {
    const trimmed = text.trim();
    if (!trimmed)
        return null;
    if (contentType.includes('text/event-stream') || trimmed.startsWith('event:')) {
        // SSE frame format: event: message\ndata: {...}\n\n
        for (const line of trimmed.split(/\r?\n/)) {
            if (line.startsWith('data:')) {
                const payload = line.slice(5).trim();
                if (!payload)
                    continue;
                try {
                    return JSON.parse(payload);
                }
                catch {
                    // Not JSON, try next line
                }
            }
        }
        return null;
    }
    try {
        return JSON.parse(trimmed);
    }
    catch {
        return null;
    }
}
